File: shared/src/iosMain/kotlin/app/player/resolver/YouTubeKitBridge.kt
package app.player.resolver

/**
 * Factory for the Swift YouTubeKit bridge, registered at app startup
 * (`YouTubeKitBridgeKt.instantiateYouTubeKitBridge = { YouTubeKitBridgeImpl() }`).
 *
 * [YouTubeKit](https://github.com/alexeichhorn/YouTubeKit) is a pure Swift package (no `@objc`),
 * so Kotlin/Native cannot see it through cinterop. The SwiftNIO bridge uses the same pattern: the
 * abstract class is declared here and subclassed in Swift (`iosApp/iosApp/YouTubeKitBridge.swift`).
 *
 * Resolution is asynchronous (a network fetch plus cipher solving). The bridge takes a callback
 * (see [YouTubeKitBridge]), and [YouTubeKitMediaResolver] wraps it in
 * `suspendCancellableCoroutine` to provide the suspend [MediaResolver.resolve] API. The Swift side
 * uses YouTubeKit's default extraction method, which is local only on iOS. YouTubeKit's remote
 * extraction server, a fallback for when YouTube changes its signature cipher, is not enabled.
 */
var instantiateYouTubeKitBridge: (() -> YouTubeKitBridge)? = null

/**
 * Bridge that the Swift side subclasses. The methods take callbacks, so the bridge does not
 * depend on Kotlin and Swift coroutine interop (which would need SKIE).
 */
abstract class YouTubeKitBridge {
    /**
     * Resolves [url] to a direct streamable URL plus best-effort metadata. Always calls
     * [completion] exactly once, from any thread:
     *  - Success: `directUrl` is not null, `title` may be null, and `durationSec` is `-1.0` when
     *    unknown (YouTubeKit's metadata struct has no duration field).
     *  - Any failure (parse, network, no playable stream): null, null and -1.0.
     */
    abstract fun resolve(
        url: String,
        completion: (directUrl: String?, title: String?, durationSec: Double) -> Unit,
    )
}

File: iosApp/iosApp/YouTubeKitBridge.swift
import Foundation
import shared
import YouTubeKit

/// Swift implementation of `YouTubeKitBridge` (declared in Kotlin's iosMain), built on the
/// [YouTubeKit](https://github.com/alexeichhorn/YouTubeKit) Swift package.
///
/// YouTubeKit is pure Swift with no `@objc` surface, so Kotlin cannot call it through cinterop.
/// The Kotlin side declares an abstract class with a callback API. This class subclasses it and
/// runs the async extraction inside a `Task`. `iOSApp.swift` registers it.
///
/// Stream selection: the highest-resolution stream with both audio and video that AVPlayer can
/// decode natively (`isNativelyPlayable`: both codecs are ones AVPlayer decodes, such as H.264
/// and AAC). Adaptive (DASH-style) audio-only and video-only streams are skipped, because the
/// players need one combined source.
class YouTubeKitBridgeImpl: YouTubeKitBridge, @unchecked Sendable {

    /// Kotlin's `(String?, String?, Double) -> Unit` boxes the Double when it crosses to
    /// Objective-C, so the Swift signature takes `KotlinDouble`, not Swift `Double`. A plain
    /// `Double` fails to compile with an override-not-found error.
    override func resolve(
        url: String,
        completion: @escaping (String?, String?, KotlinDouble) -> Void
    ) {
        let unknownDuration = KotlinDouble(value: -1.0)

        guard let videoURL = URL(string: url) else {
            completion(nil, nil, unknownDuration)
            return
        }
        Task {
            do {
                let yt = YouTube(url: videoURL)
                // A combined audio and video stream that AVPlayer decodes natively. If there is
                // none, `stream` is nil and resolution fails, so the caller uses the original URL.
                let stream = try await yt.streams
                    .filterVideoAndAudio()
                    .filter { $0.isNativelyPlayable }
                    .highestResolutionStream()

                guard let resolvedURL = stream?.url else {
                    completion(nil, nil, unknownDuration)
                    return
                }

                // Metadata fetch is best-effort: failure here doesn't abort resolution.
                let title: String? = (try? await yt.metadata)?.title
                // YouTubeMetadata has no duration, so pass -1.0 (unknown).
                completion(resolvedURL.absoluteString, title, unknownDuration)
            } catch {
                completion(nil, nil, unknownDuration)
            }
        }
    }
}
